src/common/snippets/src/runtime_configurator.cpp

```diff
@@ -272,9 +259,182 @@ void RuntimeConfigurator::update_data_offsets() const {
+}
+
+void RuntimeConfigurator::ParallelWAOptimizer::update(RuntimeConfigurator::LoopInfoRuntimeParamsMap& map,
+                                                      std::vector<VectorDims>& shapes,
+                                                      std::vector<std::vector<size_t>>& layouts,
+                                                      const std::shared_ptr<RuntimeConfig>& config,
+                                                      size_t in_num) {
+    update_split_loops_info(map);
+    update_shapes(shapes);
+    update_layouts(layouts);
+    update_config(config);
+}
+
+void RuntimeConfigurator::ParallelWAOptimizer::update_split_loops_info(LoopInfoRuntimeParamsMap& initialized_info) {
+    for (const auto& loop : loops_to_split) {
+        if (initialized_info.count(loop) == 0) {
+            loop->set_work_amount(new_m);
+            lowered::pass::InitLoops::update_data_pointer_shifts(loop);
+            initialized_info[loop] = compute_runtime_params(loop);
+        }
+    }
+}
+
+void RuntimeConfigurator::ParallelWAOptimizer::update_shapes(std::vector<VectorDims>& shapes) {
+    for (size_t i = 0; i < m_dim_idces.size(); ++i) {
+        shapes[i] = not_m_related_params.count(i)
+                        ? SplitDimensionM::unsqueeze_m_dim(shapes[i], m_dim_idces[i])
+                        : SplitDimensionM::reshape_m_dim(shapes[i], m_dim_idces[i], batch_m, new_m);
+    }
+}
+
+void RuntimeConfigurator::ParallelWAOptimizer::update_layouts(std::vector<std::vector<size_t>>& layouts) {
+    layouts = split_layouts;
+}
+
+void RuntimeConfigurator::ParallelWAOptimizer::update_config(const std::shared_ptr<RuntimeConfig>& config) {
+    *++config->master_shape.rbegin() = new_m;
+    config->master_shape.insert(config->master_shape.cbegin() + config->master_shape.size() - 2, batch_m);
+    config->tensor_rank = config->master_shape.size();
+}
+
+std::unordered_set<ExpressionPtr> RuntimeConfigurator::ParallelWAOptimizer::find_applicable_brgemms(
+    const lowered::LinearIRPtr& linear_ir) {
+    auto is_brgemm = [](const ExpressionPtr& expr) {
+        return ov::is_type<op::Brgemm>(expr->get_node());
+    };
+    auto brgemm_it = std::find_if(linear_ir->begin(), linear_ir->end(), is_brgemm);
+    std::unordered_set<ExpressionPtr> brgemms;
+    while (brgemm_it != linear_ir->end()) {
+        brgemms.insert(*brgemm_it);
+        brgemm_it = std::find_if(std::next(brgemm_it), linear_ir->end(), is_brgemm);
+    }
+    const auto& loop_manager = linear_ir->get_loop_manager();
+    auto applicable_brgemm = [&loop_manager](const ExpressionPtr& expr) {
+        const auto& loop_idces = expr->get_loop_ids();
+        if (loop_idces.empty())
+            return false;
+        const auto& outermost_loop = loop_manager->get_loop_info(loop_idces[0]);
+        // WA: the feature doesn't work in static case
+        if (!utils::is_dynamic_value(outermost_loop->get_work_amount()))
+            return false;
```

Comment:
Could you please add details in the comment? The feature doesn't work because code generation happens after runtime config update - however, LoopEnd expressions are not updated and compiled as it is.
The ticket: CVS-148805

Reply:
Just thinking out loud:
By the way, probably they're applicable also only when their work_amounts are the same (in symbolic symbols bounds).
I mean if there will be 2 Brgemms on inputs of Adds -> these brgemms M dimensions should be aligned -> then we can use this optimization. If work amounts different on runtime, we cannot use it.
Honestly, I don't know how we can resolve it now and I'm not sure - we should do it at the moment. But I'd say it's some hole for generic solution🥲 

Also there is another "hole": need to check that other Loops connected to incremented ports of loops by M exist 🤔  Just in case. (At the moment Softmax and other intermediate ops are in Loop by M but if if they missed we will get incorrect behavior)

Mu suggestion is to create the ticket and describes all limitations of the current implementation to come back and resolve them.

Reply:
The comment just before the lambda is added.

Regarding 2 weak points you described: wouldn't the pipeline fail before we call `update` in both of these situations (in case of MHA pattern of course)? If there is 2 brgemms with different M, shape inference should fail
In case if softmax is not covered with loop by M: I am not sure that we fail before `update`, but we fail regardless `ParallelWAOptimizer` is applied or not. Am I missed something?

Reply:
I just thought in MLP bounds or any another case (non-MHA) 🤔 
We don't have enough time for generic algorithm implementation, however it would be great if we have some list with current limitations to see why we can not use it as common algorithm 😃